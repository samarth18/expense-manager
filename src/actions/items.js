import uuid from "uuid";
import database from "../firebase/firebase";

// ADD_ITEM
export const addItem = item => ({
  type: "ADD_ITEM",
  item
});

export const startAddItem = (itemData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = "",
      note = "",
      quantity = 1,
      cost = 0,
      createdAt = 0,
      netAmount = 0
    } = itemData;
    const item = { description, note, quantity, cost, createdAt, netAmount };

    return database
      .ref(`users/${uid}/items`)
      .push(item)
      .then(ref => {
        dispatch(
          addItem({
            id: ref.key,
            ...item
          })
        );
      });
  };
};

// REMOVE_ITEM
export const removeItem = ({ id } = {}) => ({
  type: "REMOVE_ITEM",
  id
});

export const startRemoveItem = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/items/${id}`)
      .remove()
      .then(() => {
        dispatch(removeItem({ id }));
      });
  };
};

// EDIT_ITEM
export const editItem = (id, updates) => ({
  type: "EDIT_ITEM",
  id,
  updates
});

export const startEditItem = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/items/${id}`)
      .update(updates)
      .then(() => {
        dispatch(editItem(id, updates));
      });
  };
};

// SET_ITEMS
export const setItems = items => ({
  type: "SET_ITEMS",
  items
});

export const startSetItems = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/items`)
      .once("value")
      .then(snapshot => {
        const items = [];

        snapshot.forEach(childSnapshot => {
          items.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });

        dispatch(setItems(items));
      });
  };
};
