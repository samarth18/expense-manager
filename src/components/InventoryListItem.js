import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

const InventoryListItem = ({
  id,
  description,
  netAmount,
  createdAt,
  quantity
}) => (
  <Link className="list-item" to={`/edit/${id}`}>
    <div>
      <h3 className="list-item__title">{description}</h3>
      <div>
        <span className="list-item__sub-title">Quantity: {quantity}</span>
      </div>
      <span className="list-item__sub-title">
        {moment(createdAt).format("MMMM Do, YYYY")}
      </span>
    </div>
    <h3 className="list-item__data">{numeral(netAmount).format("$0,0.00")}</h3>
  </Link>
);

export default InventoryListItem;
