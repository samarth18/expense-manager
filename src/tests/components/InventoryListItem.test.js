import React from "react";
import { shallow } from "enzyme";
import items from "../fixtures/items";
import InventoryListItem from "../../components/InventoryListItem";

test("should render InventoryListItem correctly", () => {
  const wrapper = shallow(<InventoryListItem {...items[0]} />);
  expect(wrapper).toMatchSnapshot();
});
