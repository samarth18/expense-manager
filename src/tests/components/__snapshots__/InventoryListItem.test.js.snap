// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render InventoryListItem correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <InventoryListItem
    cost={195}
    createdAt={0}
    description="Gum"
    id="1"
    note=""
    quantity="2"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <div>
          <h3
                    className="list-item__title"
          >
                    Gum
          </h3>
          <div>
                    <span
                              className="list-item__sub-title"
                    >
                              Quantity: 
                              2
                    </span>
          </div>
          <span
                    className="list-item__sub-title"
          >
                    December 31st, 1969
          </span>
</div>,
        <h3
          className="list-item__data"
>
          $0.00
</h3>,
      ],
      "className": "list-item",
      "replace": false,
      "to": "/edit/1",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h3
              className="list-item__title"
>
              Gum
</h3>,
            <div>
              <span
                            className="list-item__sub-title"
              >
                            Quantity: 
                            2
              </span>
</div>,
            <span
              className="list-item__sub-title"
>
              December 31st, 1969
</span>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Gum",
              "className": "list-item__title",
            },
            "ref": null,
            "rendered": "Gum",
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <span
                className="list-item__sub-title"
>
                Quantity: 
                2
</span>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Quantity: ",
                  "2",
                ],
                "className": "list-item__sub-title",
              },
              "ref": null,
              "rendered": Array [
                "Quantity: ",
                "2",
              ],
              "type": "span",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "December 31st, 1969",
              "className": "list-item__sub-title",
            },
            "ref": null,
            "rendered": "December 31st, 1969",
            "type": "span",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "$0.00",
          "className": "list-item__data",
        },
        "ref": null,
        "rendered": "$0.00",
        "type": "h3",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <div>
            <h3
                        className="list-item__title"
            >
                        Gum
            </h3>
            <div>
                        <span
                                    className="list-item__sub-title"
                        >
                                    Quantity: 
                                    2
                        </span>
            </div>
            <span
                        className="list-item__sub-title"
            >
                        December 31st, 1969
            </span>
</div>,
          <h3
            className="list-item__data"
>
            $0.00
</h3>,
        ],
        "className": "list-item",
        "replace": false,
        "to": "/edit/1",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h3
                className="list-item__title"
>
                Gum
</h3>,
              <div>
                <span
                                className="list-item__sub-title"
                >
                                Quantity: 
                                2
                </span>
</div>,
              <span
                className="list-item__sub-title"
>
                December 31st, 1969
</span>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Gum",
                "className": "list-item__title",
              },
              "ref": null,
              "rendered": "Gum",
              "type": "h3",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <span
                  className="list-item__sub-title"
>
                  Quantity: 
                  2
</span>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Quantity: ",
                    "2",
                  ],
                  "className": "list-item__sub-title",
                },
                "ref": null,
                "rendered": Array [
                  "Quantity: ",
                  "2",
                ],
                "type": "span",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "December 31st, 1969",
                "className": "list-item__sub-title",
              },
              "ref": null,
              "rendered": "December 31st, 1969",
              "type": "span",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "$0.00",
            "className": "list-item__data",
          },
          "ref": null,
          "rendered": "$0.00",
          "type": "h3",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
